core/arch/atomic_exports.h

```diff
@@ -723,7 +723,254 @@ atomic_dec_becomes_zero(volatile int *var)
                                  : "r2")
 #        define SET_IF_NOT_ZERO(flag) SET_FLAG(ne, flag)
 #        define SET_IF_NOT_LESS(flag) SET_FLAG(ge, flag)
-#    endif /* X86/ARM */
+#    elif defined(DR_HOST_RISCV64)
+#        define ATOMIC_1BYTE_READ(addr_src, addr_res)               \
+            do {                                                    \
+                __asm__ __volatile__("lbu %0, 0(%1)\n\t"            \
+                                     "fence iorw,iorw"              \
+                                     :                              \
+                                     : "r"(addr_src), "r"(addr_res) \
+                                     : "memory");                   \
+            } while (0)
+#        define ATOMIC_1BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_4BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                /* Assuming little endian. */                  \
+                ASSERT(BYTE_ORDER == LITTLE_ENDIAN);           \
+                /* FIXME i#3544: test the assembly!*/          \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
```

Comment:
The sequence below is still non-atomic though, right? There are four separate stores, so it's possible for some stores to be visible before the others.

ATOMIC_4BYTE_ALIGNED_WRITE

Reply:
You're correct, this implementation is wrong. RISC-V atomic operations ISA extension (part of standard set called RV64GC ISA) only works on aligned addresses. I need to figure out how to make effectively compare and swap loop on non-aligned addresses.
That said, IIUC Aarch64 effectively only accepts aligned addresses in `ATOMIC_4BYTE_WRITE` so I could do the same?

Reply:
Sorry missed your question before. It should be okay to follow what AArch64 does yes.

Reply:
I did that. All reads/writes alias to aligned functions and contain align asserts.